hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateDagVisitor.java

```diff
@@ -426,6 +440,71 @@ public Vertex onHashJoin(JoinHashPhysicalRel rel) {
         return joinVertex;
     }
 
+    public Vertex onStreamToStreamJoin(StreamToStreamJoinPhysicalRel rel) {
+        JetJoinInfo joinInfo = rel.joinInfo(parameterMetadata);
+
+        Map<Byte, ToLongFunctionEx<JetSqlRow>> leftExtractors = new HashMap<>();
+        Map<Byte, ToLongFunctionEx<JetSqlRow>> rightExtractors = new HashMap<>();
+        List<RelDataTypeField> fieldList = rel.getRowType().getFieldList();
+
+        // map watermarked timestamps extractors to enumerated wm keys
+        for (Map.Entry<Integer, ToLongFunctionEx<JetSqlRow>> e : rel.leftTimeExtractors().entrySet()) {
+            Map<Integer, MutableByte> refByteMap = watermarkKeysAssigner.getWatermarkedFieldsKey(rel.getLeft());
+            Byte wmKey = refByteMap.get(e.getKey()).getValue();
+            leftExtractors.put(wmKey, e.getValue());
+        }
+
+        for (Map.Entry<Integer, ToLongFunctionEx<JetSqlRow>> e : rel.rightTimeExtractors().entrySet()) {
+            Map<Integer, MutableByte> refByteMap = watermarkKeysAssigner.getWatermarkedFieldsKey(rel.getRight());
+            Byte wmKey = refByteMap.get(e.getKey()).getValue();
+            rightExtractors.put(wmKey, e.getValue());
+        }
+
+        // map field descriptors to enumerated watermark keys
+        Map<Integer, MutableByte> refByteMap = watermarkKeysAssigner.getWatermarkedFieldsKey(rel);
+        Map<Byte, Map<Byte, Long>> postponeTimeMap = new HashMap<>();
+        for (Entry<Integer, Map<Integer, Long>> entry : rel.postponeTimeMap().entrySet()) {
+            Map<Byte, Long> map = new HashMap<>();
+            for (Entry<Integer, Long> innerEntry : entry.getValue().entrySet()) {
+                map.put(refByteMap.get(innerEntry.getKey()).getValue(), innerEntry.getValue());
+            }
+            postponeTimeMap.put(refByteMap.get(entry.getKey()).getValue(), map);
+        }
+
+        // fill `postponeTimeMap` with empty inner maps for unused
+        // watermarks keys to be counted by the processor as present.
+        for (MutableByte key : refByteMap.values()) {
+            postponeTimeMap.putIfAbsent(key.getValue(), emptyMap());
+        }
+
+        Vertex joinVertex = dag.newUniqueVertex(
+                "Stream-Stream Join",
+                StreamToStreamJoinP.supplier(
+                        joinInfo,
+                        leftExtractors,
+                        rightExtractors,
+                        postponeTimeMap,
+                        rel.getLeft().getRowType().getFieldCount(),
+                        rel.getRight().getRowType().getFieldCount())
+        );
+
+        // TODO: clarify it.
+        joinVertex.localParallelism(1);
+
+        // region DAG
+        Vertex leftInput = ((PhysicalRel) rel.getLeft()).accept(this);
+        Vertex rightInput = ((PhysicalRel) rel.getRight()).accept(this);
+
+        Edge left = Edge.from(leftInput).to(joinVertex, 0);
+        Edge right = Edge.from(rightInput).to(joinVertex, 1);
```

Comment:
This works only for total parallelism of 1 (1 member running 1 processor). We need to partition the data according to the equi-join key, and if there's no equi-join key, partition one side and broadcast the other.